pkg/kube/wait.go

```diff
@@ -337,49 +292,33 @@ func getPods(client kubernetes.Interface, namespace, selector string) ([]corev1.
-//
 // Modified version of https://github.com/kubernetes/kubernetes/blob/v1.14.1/pkg/kubectl/polymorphichelpers/helpers.go#L84
 func SelectorsForObject(object runtime.Object) (selector labels.Selector, err error) {
-	switch t := object.(type) {
-	case *extensionsv1beta1.ReplicaSet:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *appsv1.ReplicaSet:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *appsv1beta2.ReplicaSet:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *corev1.ReplicationController:
-		selector = labels.SelectorFromSet(t.Spec.Selector)
-	case *appsv1.StatefulSet:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *appsv1beta1.StatefulSet:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *appsv1beta2.StatefulSet:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *extensionsv1beta1.DaemonSet:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *appsv1.DaemonSet:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *appsv1beta2.DaemonSet:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *extensionsv1beta1.Deployment:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *appsv1.Deployment:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *appsv1beta1.Deployment:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *appsv1beta2.Deployment:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *batchv1.Job:
-		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
-	case *corev1.Service:
-		if t.Spec.Selector == nil || len(t.Spec.Selector) == 0 {
-			return nil, fmt.Errorf("invalid service '%s': Service is defined without a selector", t.Name)
-		}
-		selector = labels.SelectorFromSet(t.Spec.Selector)
+	obj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(object)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to convert object %T", object)
+	}
+	kind, ok, err := unstructured.NestedString(obj, "kind")
+	if err != nil || !ok {
+		return nil, fmt.Errorf("invalid object kind %T", object)
+	}
 
+	var nestedFields []string
+	switch kind {
+	case "ReplicationController", "Service":
+		nestedFields = []string{"spec", "selector"}
+	case "ReplicaSet", "StatefulSet", "DaemonSet", "Deployment":
+		nestedFields = []string{"spec", "selector", "matchLabels"}
 	default:
 		return nil, fmt.Errorf("selector for %T not implemented", object)
 	}
 
-	return selector, errors.Wrap(err, "invalid label selector")
+	labelset, ok, err := unstructured.NestedStringMap(obj, nestedFields...)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid label selector")
+	}
+	if !ok || len(labelset) == 0 {
+		return nil, fmt.Errorf("invalid object: %s is defined without a selector", kind)
+	}
+	return labels.SelectorFromSet(labelset), nil
```

Comment:
May I ask why the object is being converted to an unstructured object, then trying to fetch its metadata with internal fields? This seems very dangerous, as different versions of a particular type can use different fields for label selectors. This new code does not differentiate between the Group, Version and Kind (GVK), whereas the previous (though highly redundant) code above did.

I'd suggest leaving reverting this and leaving it as-is for the time being. This does not seem to gain us any long term benefit (in fact, it could harm us in the future).

Reply:
A very simple use case where this code could fail:

1. I registered a new CRD with GVK `myappv1.Service`
2. This new `Service` type acts like a completely different object with a completely different API than a `corev1.Service`

In this case, the old code would've identified that this object is not part of the `corev1.Service` GVK and should be ignored. In this new code, it would not, and treat it similar to the `corev1.Service` object, causing indeterminate behaviour.

Reply:
I agree with you, just leave it as it is.